t/14-isotovideo.t

```diff
@@ -238,6 +238,30 @@ subtest 'productdir variable relative/absolute' => sub {
     unlike $log, qr/assert_screen_fail_test/, 'assert screen test not scheduled';
 };
 
+subtest 'exit status from test results: no test scheduled' => sub {
+    chdir($pool_dir);
+    path(bmwqemu::STATE_FILE)->remove if -e bmwqemu::STATE_FILE;
+    path('vars.json')->remove if -e 'vars.json';
+    my $module = 'tests/failing_module';
+    my $log = combined_from { isotovideo(
+            default_opts => '--exit-status-from-test-results backend=null',
+            opts => "casedir=$data_dir/empty_test_distribution qemu_no_kvm=1 arch=i386 backend=qemu qemu=i386 novideo=1", exit_code => 100) };
+};
+
+subtest 'exit status from test results: tests scheduled' => sub {
+    chdir($pool_dir);
+    path(bmwqemu::STATE_FILE)->remove if -e bmwqemu::STATE_FILE;
+    path('vars.json')->remove if -e 'vars.json';
+    my $module_basename = 'failing_module';
+    my $module = "tests/$module_basename";
+    my $log = combined_from { isotovideo(
+            default_opts => '--exit-status-from-test-results backend=null',
+            opts => "casedir=$data_dir/tests schedule=$module qemu_no_kvm=1 arch=i386 backend=qemu qemu=i386 novideo=1", exit_code => 101) };
+    
+    like $log, qr/scheduling failing_module $module\.pm/, 'module scheduled';
+    like $log, qr/Test result \[testresults\/result\-$module_basename\.json\] fail/, 'module test result parsed';
+};
+
```

Comment:
Please carefully check the runtime of t/14-isotovideo.t as this test module often increased significantly in runtime which is hindering further development.

Reply:
Should be the fastest it can be, it only schedules the module that fails.

Pointers are welcome cuz I did saw it falling randomly locally.

Reply:
> Should be the fastest it can be, it only schedules the module that fails.

I suggest to compare runtime before and after your changes and state the numbers here for verification.

> Well, let us know how it fails because we take stability of tests very serious.

Reply:
```
os-autoinst on  master [!?] via △ v3.20.4 via 🐪 v5.26.1 via 🐍 v2.7.18 on VPN 🇩🇪 
📅2023-12-20🕙19:57:31 ➜ time prove -I. t/14-isotovideo.t 
t/14-isotovideo.t .. ok    
All tests successful.
Files=1, Tests=13,  9 wallclock secs ( 0.02 usr  0.00 sys +  6.10 cusr  4.40 csys = 10.52 CPU)
Result: PASS

real	0m9.279s
user	0m6.171s
sys	0m4.418s
```

vs this branch:
```
os-autoinst on  add_exit_on_failed_test_module [!?] via △ v3.20.4 via 🐪 v5.26.1 via 🐍 v2.7.18 on VPN 🇩🇪 took 17s632ms 
📅2023-12-20🕙19:56:19 ➜ time prove -I. t/14-isotovideo.t 
t/14-isotovideo.t .. ok    
All tests successful.
Files=1, Tests=15, 18 wallclock secs ( 0.02 usr  0.00 sys +  8.45 cusr  8.84 csys = 17.31 CPU)
Result: PASS

real	0m17.554s
user	0m8.558s
sys	0m8.848s
```

I removed one test and got:

```
real	0m13.323s
user	0m7.367s
sys	0m6.805s
```

- every `isotovideo` invocation costs like 4s (?)
- is `use OpenQA::Test::TimeLimit '20';` a limit per-subtest o per-test-file?

Reply:
on https://github.com/josegomezr/os-autoinst/commits/0df7e176 I marked the tests that run a full `isotovideo` with `# full isotovideo invocation` and placed a rough estimation on what `TimeLimit` should be. My math may be off, but so far it was fine in my local with 0 flickering tests.

Reply:
>     * every `isotovideo` invocation costs like 4s (?)

that sounds likely. That is aggravated when coverage analysis is active so we have to be very conservative regarding extending tests in t/14-isotovideo.t

How about extracting the exit status evaluation into a library for which you can use a more cost-effective unit-test?

>     * is `use OpenQA::Test::TimeLimit '20';` a limit per-subtest o per-test-file?

per test file

Reply:
Sure, and we agree that still at least one e2e test must run with the new flag to be able to properly test the feature. Right?

Reply:
Not if that makes t/14-isotovideo.t run significantly lower because I see a much higher risk that developers turn to inefficient development practices due to too long test execution runtimes rather than the risk of possibly breaking the new feature without being noticed.

Reply:
How would you cover that extra branch of the code in the main program?

Reply:
It wouldn't so that's why you should keep it as simple as possible in the main program and at best do the branching within one line so that we can keep our complete statement coverage, call it cheating, I don't mind ;)

Reply:
How's that not an inefficient development practice? :)

Reply:
Don't you agree that unit tests should each be executable within a couple of seconds? Otherwise people won't call them until they create a PR and CI runs them which means feedback only after about 0,5h and reviewers were already notified so rather wasteful for a failing "unit" test.

Reply:
> Don't you agree that unit tests should each be executable within a couple of seconds?

That's irrelevant to the test itself. You should properly test all branches when possible. If it's too slow to your taste, the test and the component should be redesigned so it can run faster.

That can be as easy as splitting E2E tests into it's own file so it can run in parallel (`prove --jobs N`) or rewrite the component so it can have more components tested in isolation.

But the approach of "let's avoid writing test cuz it runs slow" is gonna force you to go into bad practices just suffice that artificial constraint.

> Otherwise people won't call them until they create a PR and CI runs them

The contributing guide literally says:

> Every commit is checked by our CI system as soon as you create a pull request but you **should** run the os-autoinst tests locally. Checkout the build instructions for further details.

Nothing prevents a contributor from running the test.

>  and reviewers were already notified so rather wasteful for a failing "unit" test.

People can still see the code and comment over it, is not like a green test will _reveal_ the hidden source code

Reply:
> If it's too slow to your taste, the test and the component should be redesigned so it can run faster.

If you don't do it here, chances are nobody will. And I say that having complained about it myself since it can be annoying when you just had a lovely working test.

Reply:
Check the latest commit (940c7868), speed improved significantly (it didn't need a bump on the timer anymore).

`backend=null`  was being overwritten by qemu even if it's not used.